docs/import/add_disaster.js

```diff
@@ -60,14 +76,235 @@ function enableWhenReady() {
       });
 }
 
+const STATE = {
+  SAVED: 0,
+  WRITING: 1,
+  QUEUED_WRITE: 2,
+};
+Object.freeze(STATE);
+
+let state = STATE.SAVED;
+let pendingWriteCount = 0;
+
+window.onbeforeunload = () => pendingWriteCount > 0 ? true : null;
+
+/**
+ * Write the current state of {@code disasterData} to firestore.
+ * @return {?Promise<void>} Returns when finished writing or null if it just
+ * queued a write and doesn't know when that will finish.
+ */
+function updateLayersInFirestore() {
+  if (state !== STATE.SAVED) {
+    state = STATE.QUEUED_WRITE;
+    return null;
+  }
+  addLoadingElement(writeWaiterId);
+  state = STATE.WRITING;
+  pendingWriteCount++;
+  return getFirestoreRoot()
+      .collection('disaster-metadata')
+      .doc(currentDisaster)
+      .set({layers: getCurrentLayers()}, {merge: true})
+      .then(() => {
+        pendingWriteCount--;
+        const oldState = state;
+        state = STATE.SAVED;
+        switch (oldState) {
+          case STATE.WRITING:
+            loadingElementFinished(writeWaiterId);
+            return Promise.resolve();
```

Comment:
It doesn't break the covenant of the doc. We are literally returning a promise, because the chain of functions that are actually executed returns a promise. This return is inside a lambda that is passed to then(). If the Firestore fetch never finished (for instance) this line of code would never be reached, even though this function would have long ago returned its Promise.

The .then() method of a Promise returns another Promise, let's call it thenPromise. What's inside thenPromise depends on what you return in your lambda. If you return a value, thenPromise will contain that value. If you return a Promise, thenPromise will contain whatever value was inside that Promise, when your returned Promise completes.

So, inside a then, returning a resolved Promise is exactly the same as returning the value that was inside that resolved Promise.

Reply:
Ah,  because it's in a then. Sorry I misread.